File: nodejs/day02/index2.js
const fs = require("fs");

const matrixStr = fs.readFileSync("./input", "utf-8").trim();

const matrix = matrixStr
  .split("\n")
  .map(str => str.split("\t").map(n => parseInt(n, 10)));

const sum = a => a.reduce((x, y) => x + y, 0);

const isInt = n => n === ~~n;

const findDiv = r => {
  r = r.slice().sort((x, y) => x - y);
  const l = r.length;
  for (let i = 0; i < l - 1; i++) {
    const x = r[i];
    for (let j = i + 1; j < l; j++) {
      const y = r[j];
      const d = y / x;
      if (isInt(d)) {
        return d;
      }
    }
  }
};

const cs = sum(matrix.map(findDiv));

console.log(cs);

File: clojure/day02/index2.clj
(require '[clojure.string :as s])

(def parse-int #(Integer/parseInt %))

(defn split [p s] (s/split s p))

(def matrix-str (slurp "./input"))

(defn parse-row [s] (->> s
  (split #"\t")
  (map parse-int)
  (into [])
))

(def matrix (->> matrix-str
  s/split-lines
  (map parse-row)
))

(defn sum [a] (reduce + 0 a))

(defn find-div [r]
  (let [
    r' (->> r sort (into []))
    l (count r)
  ] 
  (loop [i 0 j 1] (
      let [
        x (get r' i)
        y (get r' j)
      ] 
      (cond
        (== i l) (throw (ex-info "Not found" {}))
        (== j l) (recur (+ i 1) (+ i 2))
        (== 0 (mod y x)) (/ y x)
        :else (recur i (inc j))
      )))))

(println (->> matrix
  (map find-div)
  sum
))